### DISCUSSION TITLE
map_overlap performance on large inputs

### DISCUSSION THREAD

**Discussion by emilmelnikov** [dask/dask#7404] [General]
We had a discussion with @GenevieveBuckley and @jni about our attempts to use dask arrays in [ilastik](https://www.ilastik.org/).

Our application is a 5D image viewer that also interactively applies ML algorithms to user data. One of our workflows needs to compute various filters from the input data of arbitrary size. For the purposes of this discussion, assume that the input data is blocked (HDF5, Zarr, N5). All these filters are convolutional (Gaussian, DoG etc.), so they can be computed only for blocks that fully cover the current GUI viewport. Also, in order to maintain an acceptable latency, and to reduce the amount of wasted compute, block sizes should be quite small (we use 64^3 for 3D data). This creates a performance problem if one wants to use dask arrays for all lazy computations in a pipeline, due to the large size of a dask graph for intermediate computations.

```python
import contextlib
import time

import dask
import dask.array
import zarr


def identity(x):
    return x


@contextlib.contextmanager
def measure(label):
    t0 = time.perf_counter()
    yield
    t = time.perf_counter() - t0
    print(f"{label}: {t:.9} seconds")


print(f"{dask.__version__ = }")

shape = (2**11,) * 3
chunks = (64,) * 3
src = zarr.zeros(shape, chunks=chunks, dtype="uint8")
print(f"{src.shape = }")
print(f"{src.chunks = }")
print(f"{src.nbytes / (2**30) = }")

with measure("create dask array 'a' from zarr array 'src'"):
    a = dask.array.from_zarr(src)
print(f"{len(a.__dask_graph__()) = }")

with measure("map_blocks dask array 'a' to dask array 'b'"):
    b = dask.array.map_blocks(identity, a)
print(f"{len(b.__dask_graph__()) = }")

with measure("map_overlap dask array 'a' to dask array 'c'"):
    c = dask.array.map_overlap(identity, a, depth=1, boundary="none")
print(f"{len(c.__dask_graph__()) = }")
```

Output:
```
dask.__version__ = '2021.03.0'
src.shape = (2048, 2048, 2048)
src.chunks = (64, 64, 64)
src.nbytes / (2**30) = 8.0
create dask array 'a' from zarr array 'src': 0.021725723 seconds
len(a.__dask_graph__()) = 32769
map_blocks dask array 'a' to dask array 'b': 0.000862336 seconds
len(b.__dask_graph__()) = 65537
map_overlap dask array 'a' to dask array 'c': 8.43978025 seconds
len(c.__dask_graph__()) = 994425
```

I wonder if https://github.com/dask/dask/issues/6438 is somehow related to this problem?

**Comment by mrocklin** [comment]
cc @rjzamora and @ian-r-rose who have thought about high level graphs for arrays.  One way to solve this problem would be to make a high level graph layer for map-overlap and then implement culling/dependency for that layer.

**Reply by mrocklin** [reply]
Just to add more context here, my guess is that the problem is that Dask is creating a huge graph and then throwing most of it away every time.  Recently we've started creating low-level graphs on an as-needed basis, and this has dramatically helped improve performance in some cases.  I suspect that we could apply this same treatment to `map_overlap` and achieve similar results, especially when slicing after calling `map_overlap` (which after speaking with @GenevieveBuckley , is the situation that we're in here).  

If so, the approach here may be straightforward (make high level layer class, change graph construction code to only use dependent tasks, replace old map_overlap graph construction code).  

For full high level graph benefits this may also require us to do the same treatment for slicing, which is a larger item and on the TODO list for the graph/scheduler team.  This would allow us to send a fully abstract graph to the scheduler, and skip out on a lot of overhead.

**Reply by jrbourbeau** [reply]
also cc @madsbk for visibility

**Reply by GenevieveBuckley** [reply]
Linking to discussion related to slicing: https://github.com/dask/dask/issues/5918

**Reply by rjzamora** [reply]
Once #7381 is in, it shouldn't be too difficult to pull most array creation/io under the Blockwise umbrella.  From an initial glance at the example above, it seems that a reasonable first step is to convert `from_array` and `from_zarr` into `Blockwise` layers (to avoid materializing a large graph during the creation step).

**Reply by GenevieveBuckley** [reply]
Looks like https://github.com/dask/dask/pull/7381 was merged 6 hours ago, so great timing! @rjzamora would you be willing to review a PR converting `from_array` and `from_zarr` into `Blockwise`?

**Reply by rjzamora** [reply]
> @rjzamora would you be willing to review a PR converting from_array and from_zarr into Blockwise?

Absolutely happy to help in any way I can!  Note that I have started to migrate some DataFrame-specific IO in #7415, but I haven't tackled DataFrame.from_array yet (which I would expect to require similar logic). Also, cc @ian-r-rose, who has been thinking about Blockwise IO in dask.array lately.

**Reply by GenevieveBuckley** [reply]
Started looking at how we could make a blockwise from_array this morning, and have only just realised https://github.com/dask/dask/pull/7417 was opened over the weekend (i.e. Friday for the USA). Going to take a look at that now.

**Reply by GenevieveBuckley** [reply]
Spent some more time looking at map_overlap today, it seems that task graphs get very complicated very quickly. I thought a toy example might help me:

```python
import dask.array as da

def myfunc(x):
    return x + 1

arr = da.from_array(np.ones((2, 4)).astype(int), chunks=2)
arr
```
![image](https://user-images.githubusercontent.com/30920819/111950169-07c86300-8b36-11eb-8b21-3e356bb7269e.png)

```python
result_from_blocks = arr.map_blocks(myfunc)
result_from_blocks.visualize()
```
![image](https://user-images.githubusercontent.com/30920819/111950193-11ea6180-8b36-11eb-984f-5a9473288483.png)

```python
result_with_overlap = arr.map_overlap(myfunc, depth=1, align_arrays=False, trim=False, boundary=None)
result_with_overlap.visualize()
```

![mydask](https://user-images.githubusercontent.com/30920819/111950029-d354a700-8b35-11eb-940b-ac447c563712.png)

This task graph is a bit too complicated to be very useful. I went down a bit of a rabbit hole looking for tools that would help me compare graphviz graphs more easily, but I suspect it's better to try a different approach for working out what's going on.

**Comment by mrocklin** [comment]
The blockwise work will be great in accelerating things like from_array or from_zarr, however it won't help with `map_overlap`.  Instead, we'll need to repeat the treatment that we did for blockwise on `map_overlap`.  We'll probably need to make a high level layer for `map_overlap` (or the various sub-components of that operation) to make this work well.  I would expect this to be at least a week of work. from someone.

**Reply by GenevieveBuckley** [reply]
This is a naive question perhaps, but what is the reason `overlap_internal` can't be a slice operation?

**Comment by GenevieveBuckley** [comment]
FYI @gjoseph92 & @ian-r-rose 

My colleague just pointed out a spot where re-inventing a version of map_overlap helped avoid some of the performance problems we've been talking about:
https://github.com/royerlab/dexp/blob/master/dexp/processing/utils/scatter_gather_i2i.py

**Comment by GenevieveBuckley** [comment]
@jni requested links to:

* the map_overlap PR: https://github.com/dask/dask/pull/7595
* the slicing PR: https://github.com/dask/dask/pull/7655
* this blogpost: https://blog.dask.org/2021/07/07/high-level-graphs

**Comment by GenevieveBuckley** [comment]
Ian's written a summary of the current status and next steps necessary for the high level graph work for slicing and array overlap. The gist is here: https://gist.github.com/ian-r-rose/4221ebf52f3423203640c498fb815f21
